iceoryx_examples/callbacks_in_c/ice_c_callbacks_subscriber.c

```diff
@@ -0,0 +1,165 @@
+        iox_user_trigger_trigger(heartbeat);
+        sleep_for(4000);
+    }
+    return NULL;
+}
+
+void onSampleReceivedCallback(iox_sub_t subscriber)
+{
+    const struct CounterTopic* chunk;
+    if (iox_sub_take_chunk(subscriber, (const void**)&chunk) == ChunkReceiveResult_SUCCESS)
+    {
+        iox_service_description_t serviceDescription = iox_sub_get_service_description(subscriber);
+        if (strcmp(serviceDescription.instanceString, "FrontLeft") == 0)
+        {
+            leftCache.value = *chunk;
+            leftCache.isSet = true;
+        }
+        else if (strcmp(serviceDescription.instanceString, "FrontRight") == 0)
+        {
+            rightCache.value = *chunk;
+            rightCache.isSet = true;
+        }
+        printf("received: %d\n", chunk->counter);
+    }
+
+    if (leftCache.isSet && rightCache.isSet)
+    {
+        printf("Received samples from FrontLeft and FrontRight. Sum of %d + %d = %d\n",
+               leftCache.value.counter,
+               rightCache.value.counter,
+               leftCache.value.counter + rightCache.value.counter);
+        leftCache.isSet = false;
+        rightCache.isSet = false;
+    }
+}
+
+int main()
+{
+    signal(SIGINT, sigHandler);
+    signal(SIGTERM, sigHandler);
+
+    iox_runtime_init("iox-c-callback-subscriber");
+
+    // the listener starts a background thread and the callbacks of the attached events
+    // will be called in this background thread when they are triggered
+    iox_listener_storage_t listenerStorage;
+    iox_listener_t listener = iox_listener_init(&listenerStorage);
+
+    iox_user_trigger_storage_t heartbeatStorage;
+    heartbeat = iox_user_trigger_init(&heartbeatStorage);
+
+    iox_sub_options_t options;
+    iox_sub_options_init(&options);
+    options.historyRequest = 10U;
+    options.queueCapacity = 5U;
+    options.nodeName = "iox-c-callback-subscriber-node";
+    iox_sub_storage_t subscriberLeftStorage, subscriberRightStorage;
+
+    iox_sub_t subscriberLeft = iox_sub_init(&subscriberLeftStorage, "Radar", "FrontLeft", "Counter", &options);
+    iox_sub_t subscriberRight = iox_sub_init(&subscriberRightStorage, "Radar", "FrontRight", "Counter", &options);
+
+    // send a heartbeat every 4 seconds
+#if !defined(_WIN32)
+    pthread_t heartbeatTriggerThread;
+    if (pthread_create(&heartbeatTriggerThread, NULL, cyclicHeartbeatTrigger, NULL))
+    {
+        printf("failed to create thread\n");
+        return -1;
+    }
+#endif
+
+    // attach everything to the listener, from here one the callbacks are called when an event occurs
+    iox_listener_attach_user_trigger_event(listener, heartbeat, &heartbeatCallback);
+    iox_listener_attach_subscriber_event(listener, subscriberLeft, SubscriberEvent_HAS_DATA, &onSampleReceivedCallback);
+    iox_listener_attach_subscriber_event(
+        listener, subscriberRight, SubscriberEvent_HAS_DATA, &onSampleReceivedCallback);
+
+    // wait until someone presses CTRL+c
+    while (keepRunning)
```

Comment:
We do not have a portable semaphore in C or why the difference to the C++ example?

Reply:
Here we have the following problem:
1. Our platform headers are only available for the utils and C++
2. using `include <semaphore.h>` will lead to an example which does not work in Windows and on Mac OS

I thought this would be the best compromise.